### ISSUE TITLE
remove all `torch.no_grad():`

### ISSUE THREAD

**Issue by mli** [d2l-ai/d2l-en#1741]
I don't think it is necessary, but there are still multiple notebooks containing this sentence:

```
(base) $ ack no_grad
chapter_linear-networks/linear-regression-scratch.md
308:    with torch.no_grad():
387:    with torch.no_grad():


chapter_convolutional-neural-networks/lenet.md
355:            with torch.no_grad():

chapter_computational-performance/multiple-gpus.md
318:    with torch.no_grad():

chapter_appendix-mathematics-for-deep-learning/maximum-likelihood.md
161:    with torch.no_grad():

chapter_recurrent-modern/seq2seq.md
594:            with torch.no_grad():

chapter_optimization/adam.md
79:        with torch.no_grad():
175:        with torch.no_grad():

chapter_optimization/adagrad.md
145:        with torch.no_grad():

chapter_optimization/momentum.md
183:        with torch.no_grad():

chapter_optimization/adadelta.md
77:        with torch.no_grad():

chapter_optimization/lr-scheduler.md
117:            with torch.no_grad():

chapter_optimization/rmsprop.md
116:        with torch.no_grad():
```

**Comment by AnirudhDagar** [comment]
@mli @astonzhang `with torch.no_grad():` just makes sure that unnecessary gradient computation is avoided. It is not necessary but a good practice since you reduce memory usage. I still feel we should keep it, and in fact, use it whenever there is no need to track gradients.

**Comment by AnirudhDagar** [comment]
Closing this for now, please feel free to re-open if a discussion is required in favour of removing `torch.no_grad`.
